File: 0003-longest-substring-without-repeating-characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int start = 0;
        int longestSubstringLen = 0;
        vector<int> charInd(256, -1);

        for(int i=0; i<s.length(); i++){
            char ch = s[i];

            if(charInd[ch] == -1){
                charInd[ch] = i;
            }else{
                int firstOccInd = charInd[ch];
                if(firstOccInd < start){
                    // do nothing
                }else{
                    start = firstOccInd+1;
                }

                charInd[ch] = i;
            }

            longestSubstringLen = max(longestSubstringLen, i-start+1);
        }

        return longestSubstringLen;
    }
};

File: 0003-longest-substring-without-repeating-characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        // my 1st time submited code (apna college + my logic)
        int n = s.length();
        if (n == 0)
            return 0;

        int indexOfChar[] = new int[256];
        for (int i = 0; i < 256; i++)
            indexOfChar[i] = -1;

        int maxLen = 0, start = -1;
        for (int i = 0; i < n; i++) {

            if (indexOfChar[s.charAt(i)] > -1 ){
                if( indexOfChar[s.charAt(i)] >= start) {
                    // maxLen = Math.max(i - start, maxLen);
                    start = indexOfChar[s.charAt(i)] ;
                }else{
                    maxLen = Math.max(i - start, maxLen);
                }
            } else {
                maxLen = Math.max(i - start, maxLen);
            }
            indexOfChar[s.charAt(i)] = i;

        }
        // if(start == -1)return n;

        return maxLen;
    }
}
